sound/soc/sof/topology.c

```diff
@@ -3257,88 +3283,101 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
 			dev_err(scomp->dev, "error: unexpected DAI config count %d!\n",
 				le32_to_cpu(cfg->num_hw_configs));
 			return -EINVAL;
 		}
+		number_configs = 1;
 	} else {
 		dev_dbg(scomp->dev, "tplg: %d hw_configs found, default id: %d!\n",
 			cfg->num_hw_configs, le32_to_cpu(cfg->default_hw_config_id));
 
-		for (i = 0; i < num_hw_configs; i++) {
-			if (cfg->hw_config[i].id == cfg->default_hw_config_id)
+		for (current_config = 0; i < number_configs; current_config++) {
+			if (hw_config[current_config].id == cfg->default_hw_config_id)
 				break;
 		}
 
-		if (i == num_hw_configs) {
+		if (current_config == number_configs) {
 			dev_err(scomp->dev, "error: default hw_config id: %d not found!\n",
 				le32_to_cpu(cfg->default_hw_config_id));
 			return -EINVAL;
 		}
 	}
 
-	/* configure dai IPC message */
-	hw_config = &cfg->hw_config[i];
+	/* reserve memory for all hw configs, eventually freed by widget */
+	config = kzalloc(number_configs * sizeof(*config), GFP_KERNEL);
+	if (!config)
+		return -ENOMEM;
 
-	config.hdr.cmd = SOF_IPC_GLB_DAI_MSG | SOF_IPC_DAI_CONFIG;
-	config.format = le32_to_cpu(hw_config->fmt);
+	/* get any common DAI tokens */
+	ret = sof_parse_tokens(scomp, config, dai_link_tokens,
+			       ARRAY_SIZE(dai_link_tokens), private->array,
+			       le32_to_cpu(private->size));
+	if (ret != 0) {
+		dev_err(scomp->dev, "error: parse link tokens failed %d\n",
+			le32_to_cpu(private->size));
+		goto err;
+	}
+
+	config->hdr.cmd = SOF_IPC_GLB_DAI_MSG | SOF_IPC_DAI_CONFIG;
+	config->format = le32_to_cpu(hw_config->fmt);
+
+	/* copy common data to all config ipc structs */
+	for (i = 1; i < number_configs; i++) {
+		config[i].hdr.cmd = SOF_IPC_GLB_DAI_MSG | SOF_IPC_DAI_CONFIG;
+		config[i].format = hw_config[i].fmt;
+		config[i].type = config[i - 1].type;
+		config[i].dai_index = config[i - 1].dai_index;
+	}
 
 	/* now load DAI specific data and send IPC - type comes from token */
-	switch (config.type) {
+	switch (config->type) {
 	case SOF_DAI_INTEL_SSP:
 		ret = sof_link_ssp_load(scomp, index, link, cfg, hw_config,
-					&config);
+					config);
 		break;
 	case SOF_DAI_INTEL_DMIC:
 		ret = sof_link_dmic_load(scomp, index, link, cfg, hw_config,
-					 &config);
+					 config);
 		break;
 	case SOF_DAI_INTEL_HDA:
 		ret = sof_link_hda_load(scomp, index, link, cfg, hw_config,
-					&config);
+					config);
 		break;
 	case SOF_DAI_INTEL_ALH:
 		ret = sof_link_alh_load(scomp, index, link, cfg, hw_config,
-					&config);
+					config);
 		break;
 	case SOF_DAI_IMX_SAI:
 		ret = sof_link_sai_load(scomp, index, link, cfg, hw_config,
-					&config);
+					config);
 		break;
 	case SOF_DAI_IMX_ESAI:
 		ret = sof_link_esai_load(scomp, index, link, cfg, hw_config,
-					 &config);
+					 config);
 		break;
 	default:
 		dev_err(scomp->dev, "error: invalid DAI type %d\n",
-			config.type);
+			config->type);
 		ret = -EINVAL;
 		break;
 	}
 	if (ret < 0)
-		return ret;
+		goto err;
 
 	return 0;
+
+err:
+	kfree(config);
```

Comment:
you have to free `config` in the success case too some time later, or use `devm_kcalloc()`

Reply:
this is freed in L2594 in widget free. config is set to dai_config, it was previously set with kmemdup, so I think I'm not changing the functionality? I'm just reserving the area in an earlier stage than before.